File: clients/php/src/Result/GrantResult.php
<?php

declare(strict_types=1);

namespace Priostack\Acn\Result;

/**
 * Result of {@see \Priostack\Acn\Client::grantAccess()} and
 * {@see \Priostack\Acn\Client::approveRequest()}.
 *
 * Keep the {@see $capabilityRef} to later revoke the grant.
 */
final class GrantResult
{
    /**
     * @param list<string>         $effectiveRights
     * @param array<string, mixed> $raw
     */
    public function __construct(
        public readonly string $capabilityRef,
        public readonly string $subject,
        public readonly string $resource,
        public readonly array $effectiveRights,
        public readonly array $raw = [],
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromData(array $data): self
    {
        return new self(
            capabilityRef: (string) ($data['capabilityRef'] ?? ''),
            subject: (string) ($data['subject'] ?? ''),
            resource: (string) ($data['resource'] ?? ''),
            effectiveRights: array_values((array) ($data['effectiveRights'] ?? [])),
            raw: $data,
        );
    }
}

File: clients/java/src/main/java/com/priostack/acn/GrantResult.java
package com.priostack.acn;

import com.fasterxml.jackson.databind.JsonNode;
import java.util.List;

/** Result of {@link ACNClient#grantAccess} / {@link ACNClient#approveRequest}. */
public final class GrantResult {

    private final String capabilityRef;
    private final String subject;
    private final String resource;
    private final List<String> effectiveRights;
    private final JsonNode raw;

    public GrantResult(
            String capabilityRef,
            String subject,
            String resource,
            List<String> effectiveRights,
            JsonNode raw) {
        this.capabilityRef = capabilityRef;
        this.subject = subject;
        this.resource = resource;
        this.effectiveRights = effectiveRights;
        this.raw = raw;
    }

    /** The capability reference; pass it to {@link ACNClient#revokeAccess} to revoke. */
    public String capabilityRef() {
        return capabilityRef;
    }

    public String subject() {
        return subject;
    }

    public String resource() {
        return resource;
    }

    public List<String> effectiveRights() {
        return effectiveRights;
    }

    public JsonNode raw() {
        return raw;
    }

    @Override
    public String toString() {
        return "GrantResult{capabilityRef="
                + capabilityRef
                + ", subject="
                + subject
                + ", resource="
                + resource
                + ", effectiveRights="
                + effectiveRights
                + '}';
    }
}
